### ISSUE TITLE
Problem with video stream

### ISSUE THREAD

**Issue by Calst54** [chillibasket/walle-replica#30]
Hi chillibasket
I built the wall-e with a raspberry4 and added the raspberry camera module 2. Raspberry is runnimng on OS 64bit. now everything motors, servos run fine on the web page. all files app.py, mjpgstreamer.sh are without modification except auto start for arduino and streamer. with the new OS the camera does not have to be enabled. that's automativ and the cameara is working when I run rpicam-hello -t- 0
my problem is that i have no video stream on the web server page. the auto start seems to work since i get the "end stream" message on the web page.
i tried several things but so far nothing seems to solve the problem. can you help?
as i am rather mechanical engineer I have no idea where to search...

**Comment by chillibasket** [comment]
Hi! Thanks for letting me know that the camera is enable by default now, I'll update the instructions to mention that.

To double check that the video stream is actually running, you can open the stream separately in your browser by using the the same IP address as the web-interface, but with ":8080" at the end instead of ":5000". If that is running, then the issue is that the web-interface is not loading the stream correctly from this address. If it is not running, then most likely the configurations within the `~/mjpg-streamer.sh` file are incorrect for your camera.

**Comment by Calst54** [comment]
![IMG_3133](https://github.com/chillibasket/walle-replica/assets/128734297/96a2d831-a48f-4197-8047-80173df64842)
Hello
Thank's for your fast response. I tried to run 192.168.0.168:8080 separate and in parallel to the :5000 but I get mo image. I tried to attach a picture of "my" wall-e in a military version. (olive, matte PLA) and it has the box of the camera in front of the robot. the camera is really the last step to have a fully working wall-e. I am anxious to get this working...
regards

**Comment by chillibasket** [comment]
Great job on your Wall-E, I like the picture! The system uses "mjpg-streamer" in the background to get the video from the camera and show the live stream. It looks like the author of that tool hasn't updated it for the past 3 years, so it no longer runs properly on the newest RPi OS version (there were a lot of changes to how the camera is managed). It may be possible that the problem is fixed in one of the community forks of the library, for example in https://github.com/neuralassembly/mjpg-streamer/tree/master. There is also an issue in the library Github page which suggests they have found a potential workaround: https://github.com/jacksonliam/mjpg-streamer/issues/397

Unfortunately, I don't have the new raspberry pi to test different options with, so there is not too much I can do at the moment to solve the issue. If you manage to get the stream working by suggestions in the links above, let me know so that I can update the tutorial.

**Comment by Calst54** [comment]
Glad you like it! I spent quite some time to get there. I made some parts to organize the electrical part and I am using  lithium battery from motorbikes with 24Wh. In this setup the wall-e has sufficient energy to run for hours. Instead of the oled display I used a small voltmeter wich shows dircetly the voltage of the battery. And under the voltmeter is a HDMI port so i can use wall-e as backup computer. Once it is finished I will made the files for the electric and the port available. 
**maybe you have a hint how I could reduce sensitivity of the XBOX controller for the motors.** I tried to reduce acceleration but this had negative effect that the brake are affected as well.

I went trough all the steps of the WEB installation again to see if I had made some mistake. I only get a Error message on step 4 when I install Flask:  error: externally-managed-environment. otherwise everything runs fine and when  activating de WEB page nor error message appears. I assue the files are at the correct place.

I will try the links you suggested and let you know if anything is successfull otherwise the project may have to stop here and wait if there is a solution later. I guess people will buy new RPi and install the actuel OS and they may encounter the same problem & perhaps someone has the solution for me.

**Comment by w45354** [comment]
yes, Flask is already installed in new RPI, you can check the list by typing sudo pip3 list. BTW, how do you find the serial port of arduino at the step 9a9B?

**Comment by Calst54** [comment]
Indeed there is a whole list and Flask 2.2.2 is in it. 
As I am far away from a SW specialist I have no idea what step 9a9B is or how I get there. Is this some debug mode?

**Comment by Calst54** [comment]
for a change. picure shows wall-e with his friends. two robots that I made with ESP32 CAM and one servo to turn the head.
![IMG_3138](https://github.com/chillibasket/walle-replica/assets/128734297/d0e42a68-4e76-4f75-8e64-cbafd296a0af)

**Comment by w45354** [comment]
My USB camera works well. Maybe you can try another cam.

**Comment by Calst54** [comment]
I have ordered a 2nd camera but i don't think it is a camera problem since i get cristal a clear picture with rpicam.
question: does any of those inteface settings needs to be enabled?
![IMG_3139](https://github.com/chillibasket/walle-replica/assets/128734297/323aa846-f63c-4d1f-b290-7876e5bf1820)

**Comment by w45354** [comment]
I enabled SSH, I2C, and serial console

**Comment by Calst54** [comment]
no success. also after doing a complete new installation I get everything working but no stream. As I am unable to find the problem i will abandon the project and have a blind walle!
Thanks for you patience.

**Comment by w45354** [comment]
what a pity! your design is so great. would you mind if share your design?

**Comment by dkrey** [comment]
I just got it working with a RPi v3 Camera, but it should work with a v2 camera just fine. 
My setup is a RPi 4 2GB running Raspi OS /Debian Bookwork / 12.4

Instead of mjpg streamer switch over to camera-streamer
https://github.com/ayufan/camera-streamer

The dpkg packages didn't work, but compiling it was straight forward. 

For the webinterface I only added a quick hack for the moment.
in app.py line 23  set `autoStartCamera = True`
in line 42 set `streaming = 1`

This should already work once you've got the systemd service of camera-streamer up and running.
As another precaution I commented out most of the code from the function `def onoff_streamer():`

**Comment by Calst54** [comment]
@chillibasket: you probably like my new version. redesigned the door to fit a 4.3" display from waveshare and put the main switch in the front panel. I can stll use the HDMI port for a 2nd display. that works just perfect and looks very sophisticated. sure i will my files availabe but i still doing some design.
![IMG_3144](https://github.com/chillibasket/walle-replica/assets/128734297/72c77342-b2f1-4d16-a447-2ac9cf3ffd3e)
![IMG_3143](https://github.com/chillibasket/walle-replica/assets/128734297/ecc679f8-f717-473b-9d0d-81d979ef7888)

@dkrey : thanks for your comment. I have a second RPi with cam on the side where I can test and hook it to the andruino of walle.

**Comment by Calst54** [comment]
@dkrey : when executing he system configuration raspi ask for a password (but not mine...) and then i get the error: failed to enable unit, 
any idea? 
![IMG_3147](https://github.com/chillibasket/walle-replica/assets/128734297/702fb57e-e289-425c-9428-b55e9c54f912)

**Comment by dkrey** [comment]
@Calst54 Linux is asking for your admin password, because systemctl commands have to be run as root. But "invalid utf8" looks a bit strange on its own. Have you tried it via sudo?
`sudo systemctl enable camera-streamer`

Beside that  running camera-streamer from the shell does work in user-context.

This should run and provide a camera-stream without having to enter a password.
This is the command for the rpi v3 camera, for v2 it differs a bit.

```
/usr/local/bin/camera-streamer \
  --camera-path=/base/soc/i2c0mux/i2c@1/imx708@1a \
  --camera-type=libcamera \
  --camera-format=YUYV \
  --camera-width=2304 --camera-height=1296 \
  --camera-fps=30 \
  --camera-nbufs=2 \
  --camera-snapshot.height=1080 \
  --camera-video.height=720 \
  --camera-stream.height=480 \
  --camera-options=AfMode=2 \
  --camera-options=AfRange=2 \
  --http-listen=0.0.0.0 \
  --http-port=8080 \
  --rtsp-port
```

**Comment by Calst54** [comment]
i have no idea of the admin password. i took a new card and put the OS with imager on it. in the installation i set the PW to 1234 (its only a test card) and the host to walle. after that i installed the web server acc to the instructions form chillibasked and then made the changes you suggested. after this i used the link to the camera-streamer and this installs e file camera-streamer-generic_0.2.8.bookworm_armhf.deb and probably a lot of other files. then i used the recommended precompiled debian package for the installation. first part was ok and secon part for system configuration resulted in the error message.
but not mine...but not  
when i run the sudo as above i get: failed to enable unit: Unit file camera-streamer.service does not exist

when tried to used the direct way it can't finde the file camera-streamer. is probably in another path...

**Comment by Calst54** [comment]
found the file in /bin and change the path but now i have the messsage the libcamera is not supported..

so far the only way to get a picture of my system ist with rpicamera-hello -t 0 
so i guess the RPi and the camera are working ok

**Comment by dkrey** [comment]
Have uninstalled the dpkg package again? The libcamera message is exactly what I got when I used the .deb
Also most of the examples are missing from the deb packages.

**Comment by Calst54** [comment]
would it not be better or simpler to use the raspberry command rpicam-vid which is directly streaming a video rather than use an external tool? 
with rpicam-vid -t 0 I have the video stream of the camera on the display of the RPi and this could be sent to the app.py
can you help?

**Comment by chillibasket** [comment]
That may be an option. According to the [documentation](https://github.com/raspberrypi/documentation/blob/develop/documentation/asciidoc/computers/camera/rpicam_vid.adoc), the following command could be used to start a stream: `rpicam-vid -t 0 --inline --listen -o tcp://0.0.0.0:8080`. You can try running this command and then press the start webcam button on the web interface. If it works, then it should appear in the web interface video panel. 

If it works, the web interface script which is used to start the camera can be replaced with this simple command.

**Comment by dkrey** [comment]
I gave it a quick shot, but the process only ran a second or two. But I'm keeping my fingers crossed that @Calst54 has more success.
In the meantime I stay with camera-streamer ;)


Here's my error log btw.
```
wall-e@wall-e:~ $ rpicam-vid -t 0 --inline --listen -o tcp://0.0.0.0:8080
[0:04:43.964998375] [2949]  INFO Camera camera_manager.cpp:284 libcamera v0.1.0+118-563cd78e
[0:04:44.017831887] [2955]  WARN RPiSdn sdn.cpp:39 Using legacy SDN tuning - please consider moving SDN inside rpi.denoise
[0:04:44.020263149] [2955]  INFO RPI vc4.cpp:444 Registered camera /base/soc/i2c0mux/i2c@1/imx708@1a to Unicam device /dev/media0 and ISP device /dev/media2
[0:04:44.020340928] [2955]  INFO RPI pipeline_base.cpp:1142 Using configuration file '/usr/share/libcamera/pipeline/rpi/vc4/rpi_apps.yaml'
Preview window unavailable
Mode selection for 640:480:12:P
    SRGGB10_CSI2P,1536x864/0 - Score: 1486.67
    SRGGB10_CSI2P,2304x1296/0 - Score: 1786.67
    SRGGB10_CSI2P,4608x2592/0 - Score: 2686.67
[0:04:50.108312497] [2949]  INFO Camera camera.cpp:1183 configuring streams: (0) 640x480-YUV420 (1) 1536x864-SBGGR10_CSI2P
[0:04:50.109440858] [2955]  INFO RPI vc4.cpp:608 Sensor: /base/soc/i2c0mux/i2c@1/imx708@1a - Selected sensor format: 1536x864-SBGGR10_1X10 - Selected unicam format: 1536x864-pBAA
#0 (0.00 fps) exp 32876.00 ag 5.99 dg 1.01
#1 (30.01 fps) exp 32876.00 ag 5.99 dg 1.01
terminate called after throwing an instance of 'std::runtime_error'
  what():  failed to send data on socket
Aborted
```

**Comment by RoBurr** [comment]
Hi I'm running in the same problem(s) camera stream does not work, As well autostart is not working with Raspi Pi 4 :-(

**Comment by dkrey** [comment]
@RoBurr
Autostart most likely doesn't work because of a missing env variable.
Have a look at my comment here:
https://github.com/chillibasket/walle-replica/issues/29#issuecomment-1913716764

I never got mjpg-streamer running, but as noted above the raspi cam v3 works just fine with camera-streamer. You just need to compile it, since the packages are broken.

**Comment by RoBurr** [comment]
Thank you for your quick support, I will organize a v3 camera and try to fix both issues

**Comment by Calst54** [comment]
@chillibasket 
because i don't know where to put the string I tried it with the terminal and then opened Firefox and tried to access with :8080
on the RPi the preview window starts but on Browswer I see the following:
![grafik](https://github.com/chillibasket/walle-replica/assets/128734297/b811ded6-929d-413b-997f-49844d147305)

opening Chrome starts the preview window for a second ond then stops the process on the RPi

i assume some data is transmitted but in a format that the browser can't read

**Comment by chillibasket** [comment]
Hi @Calst54, another Wall-E builder has come up with an installation guide using a different camera streaming library which looks promising here: https://wired.chillibasket.com/3d-printed-wall-e/#comment-1277

Since this comment thread has gone a little bit off topic I will be closing it and have opened a new Issue #32 which is aimed just at fixing the camera issue. However, feel free to continue adding comments to this closed thread if you do any more testing (if required, they can be referenced on the new Issue).
